Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -177,6 +177,186 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    {
+        // outPositions holds the iterators to results of the bulk queries.
+        // In the case of the bilinear sampler, we'll be making 4 queries per
+        // input position.
+        outPositions.resize(inPositions.size() * 4);
+        outTerrainExists.resize(inPositions.size() * 4);
+        indexStepSize = 4;
+    }
+    else
+    {
+        outPositions.resize(inPositions.size());
+        outTerrainExists.resize(inPositions.size());
+    }
+
+    GenerateQueryPositions(inPositions, outPositions, sampler, indexStepSize);
+
+    AZ::Aabb bounds;
+    AZ::EntityId prevAreaId = FindBestAreaEntityAtPosition(outPositions[0].GetX(), outPositions[0].GetY(), bounds);
+    
+    // We use a sliding window here and update the window end for each
+    // position that falls in the same area as the previous positions. This consumes lesser memory
+    // than sorting the points into separate lists and handling putting them back together.
+    // This may be sub optimal if the points are randomly distributed in the list as opposed
+    // to points in the same area id being close to each other.
+    size_t windowStart = 0;
+    size_t windowEnd = 0;
+    const size_t numPositions = outPositions.size();
+    for(int i = 1; i < numPositions; i++)
+    {
+        AZ::EntityId areaId = FindBestAreaEntityAtPosition(outPositions[i].GetX(), outPositions[i].GetY(), bounds);
+        bool queryHeights = false;
+        if (areaId == prevAreaId)
+        {
+            // Update window end to current position.
+            // If it's the last position, submit the query.
+            windowEnd = i;
+            if (windowEnd == numPositions - 1)
+            {
+                queryHeights = true;
+            }
+        }
+        else
+        {
+            queryHeights = true;
+        }
+
+        if (queryHeights)
+        {
+            size_t spanLength = (windowEnd - windowStart) + 1;
+            Terrain::TerrainAreaHeightRequestBus::Event(prevAreaId,
+                &Terrain::TerrainAreaHeightRequestBus::Events::GetHeights,
+                AZStd::span<AZ::Vector3>(outPositions.begin() + windowStart, spanLength), AZStd::span<bool>(outTerrainExists.begin() + windowStart, spanLength));
+            
+            // Sometimes, a position generated for a bilinear sample may be outside
+            // of all areas. This won't be caught by our world bounds check since we
+            // only do them on input positions and not the generated ones.
+            while (areaId == AZ::EntityId())
+            {
+                // Out of bounds. Set height to world min.
+                outPositions[i].SetZ(minHeight);
+                i++;
+                if (i < numPositions)
+                {
+                    areaId = FindBestAreaEntityAtPosition(outPositions[i].GetX(), outPositions[i].GetY(), bounds);
+                }
+                else
+                {
+                    break;
+                }
+            }
```

Comment:
I found this case(in the ProcessNormalsFromList unit test) where some points generated for bilinear samples were outside all areas. In this particular unit test scenario, the world size was a lot larger than the one registered area. So the point was actually within world bounds but was outside of all areas. Is this a case that could come up in real user scenario?

Reply:
We may be able to get rid of this. See my previous comment.